pyccel/parser/semantic.py

```diff
@@ -1214,8 +1214,9 @@ def _annotate_the_called_function_def(self, old_func, function_call_args=None):
         # Set the Scope to the FunctionDef's parent Scope and annotate the old_func
         self._scope = sc
         self._visit_FunctionDef(old_func, function_call_args=function_call_args)
+        new_name = self.scope.get_expected_name(old_func.name)
```

Comment:
`get_expected_name` does  not seem to catch this kind of conflicts.

```
def f():
    do_0001 = 5
    return g() + do() + do_0001

def g():
    return 2

def do():
    return 4

a = f()
```

Reply:
This has improved after fixing the issue above. It is now flagged as a type redefinition. This occurs because the scope of `do_0001` is smaller than the scope of `do()` so `do_0001` doesn't get renamed in the syntactic stage.
There is a discussion to be had about whether we should be more enthusiastic about avoiding this kind of mix-up but it is so rare that it is probably not worth the additional code that would be needed to fix it. What do you think?

I have clarified the error that is raised though to help the user understand what the issue is:
![image](https://github.com/pyccel/pyccel/assets/25363473/f4efe361-08ae-4aed-af2d-4db2bb9331db)

[cbe8f67](https://github.com/pyccel/pyccel/pull/1914/commits/cbe8f67e03c497493c12c8ec97ae556f36d62af5)
[d98aab6](https://github.com/pyccel/pyccel/pull/1914/commits/d98aab65ff19827f2b7b1d637584fa6cd07a1a24)

Reply:
I think we should think of a systematic way to deal with scope and name collisions to avoid dealing with these corner cases. As for the fix, I wonder if you had in mind to prevent the following:
```
def f():
    y = 2.5
    return y

y = 5
a = f()
```
This is caught as an error, but I don't think it should.

Reply:
> I think we should think of a systematic way to deal with scope and name collisions to avoid dealing with these corner cases.

We have one :sweat_smile: . It is `Scope` which is found [here](https://github.com/pyccel/pyccel/blob/devel/pyccel/parser/scope.py) and documented [here](https://github.com/pyccel/pyccel/blob/devel/developer_docs/scope.md). That's why this PR is so small. It's just that it's quite difficult to catch when people forget to use it.

This is why I asked @smazouz42 to add a test for this in his PR as it is more or less the only way to check if the Scope is being used correctly.



> I wonder if you had in mind to prevent the following

Yes exactly this must not be allowed in my opinion.

> This is caught as an error, but I don't think it should.

Then it seems that we don't agree here :sweat_smile: 

I have in mind your case, but also this case:
```python
y = 3.0

def f():
    print(y)
```
Here the `y` that is local to `f` **is** the global `y` and it is quite difficult to distinguish between the local y and the global y.

At some point I would also like to support:
```python
y = 3.0

def f():
    global y
    y = 4.0
    print(y)

print(y)
f()
print(y)
```

In this case we want to see the error displayed currently if we do:
```python
y = 3.0

def f():
    global y
    y = True
    print(y)

print(y)
f()
print(y)
```

For me the simplest way to implement `global` is to simply forbid having a local object which is in collision with a global object.

You may disagree but this is then beyond the scope of this PR as this is a choice that has been made for that case and is unrelated to the bug being fixed here.

Reply:
Got it!